paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/PartitionMarkDoneAction.java

```diff
@@ -0,0 +1,122 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.action;
+
+import org.apache.paimon.CoreOptions;
+import org.apache.paimon.flink.sink.partition.AddDonePartitionAction;
+import org.apache.paimon.flink.sink.partition.SuccessFileMarkDoneAction;
+import org.apache.paimon.metastore.MetastoreClient;
+import org.apache.paimon.options.Options;
+import org.apache.paimon.table.FileStoreTable;
+
+import org.apache.flink.table.utils.PartitionPathUtils;
+
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.paimon.CoreOptions.METASTORE_PARTITIONED_TABLE;
+import static org.apache.paimon.flink.FlinkConnectorOptions.PARTITION_MARK_DONE_ACTION;
+import static org.apache.paimon.utils.Preconditions.checkArgument;
+import static org.apache.paimon.utils.Preconditions.checkNotNull;
+
+/** Table partition mark done action for Flink. */
+public class PartitionMarkDoneAction extends TableActionBase {
+
+    private final FileStoreTable fileStoreTable;
+    private final List<LinkedHashMap<String, String>> partitions;
+    private final CoreOptions coreOptions;
+    private final Options options;
+
+    public PartitionMarkDoneAction(
+            String warehouse,
+            String databaseName,
+            String tableName,
+            List<LinkedHashMap<String, String>> partitions,
+            Map<String, String> catalogConfig) {
+        super(warehouse, databaseName, tableName, catalogConfig);
+        if (!(table instanceof FileStoreTable)) {
+            throw new UnsupportedOperationException(
+                    String.format(
+                            "Only FileStoreTable supports partition_mark_done action. The table type is '%s'.",
+                            table.getClass().getName()));
+        }
+
+        this.fileStoreTable = (FileStoreTable) table;
+        this.partitions = partitions;
+        this.coreOptions = fileStoreTable.coreOptions();
+        this.options = coreOptions.toConfiguration();
+    }
+
+    @Override
+    public void run() throws Exception {
+        List<org.apache.paimon.flink.sink.partition.PartitionMarkDoneAction> actions =
+                getPartitionMarkDoneActions();
+        for (LinkedHashMap<String, String> partition : partitions) {
+            try {
+                for (org.apache.paimon.flink.sink.partition.PartitionMarkDoneAction action :
+                        actions) {
+                    action.markDone(PartitionPathUtils.generatePartitionPath(partition));
```

Comment:
Can we introduce a `PartitionPathUtils.generatePartitionPath(Map<String, String> partitionSpec, RowType partitionType)`? `LinkedHashMap` can be generated in it.

Reply:
done